### ISSUE TITLE
Is the inlang-ide-extension-plugin required? 

### ISSUE THREAD

**Issue by samuelstroschein** [inlang/inlang#758]
## Problem

The ide-extension plugin seems redundant and seems to confuse potential users. 

1. The ide extension plugin only defines the [ide extension config schema](https://github.com/inlang/inlang/tree/main/source-code/ide-extension-plugin). 
2. We expect plugins to write the config, making the ide extension plugin redundant in almost all (every?) cases.
3. Users seem to add `sdkPlugin` and `ideExtension` plugin simultaneously, believing that they need both.   

## Proposal

To be discussed in this issue. 

## Additional information

The ide extension config has been part of `@inlang/core` before. https://github.com/inlang/inlang/issues/618 raised a discussion to move the config outside of core. While the arguments in the issue are sound, the current solution seems far from ideal.

The root cause seems to be the fact that the ide extension does not have properties that are going to be used by other apps BUT the config properties are expected to be set by plugins. The expectations that config properties are set by plugins are in stark contrast to the SDK-JS. 

So, we are facing a kind of new situation where: 

- The ide extension config is an "app" in the inlang ecosystem. 
- The app config properties are expected to be set by plugins. 
- The app config schema should ideally not be included in @inlang/core, see https://github.com/inlang/inlang/issues/618
- But the config schema should also not be a dedicated plugin itself.

**Comment by samuelstroschein** [comment]
Let's discuss: 

1. Is this issue "sound"/do you agree that the plugin config schema situation is far from ideal?
2. What can we come up with to solve this issue?

**Comment by felixhaeberle** [comment]
@samuelstroschein Thanks for bringing this up! 🙏

As a user, I would like to install the ide extension and have it work without anything further do. 
But as we know, this is not possible.

Now we have different possibilities from a technical perspective:

- let there be default settings and tell the people which they are (no plugin required, opinionated behavior)

- let there be a plugin (opinionated behavior as default, main entry point for configuration) -> if the user decides only to use 
*the ide extension and nothing else from inlang, an ide-extension-plugin is needed – otherwise this can be set by any other plugin.*

- let this be not a plugin setting but a native vs code setting which the extension can expose 
*now we have to sync between inlang config plugin and vs code setting, probably not the best way, but convenient for the user to go to the vs code settings to configure our extension, limited ability there to write complete code blocks such as messageReferenceMatchers*

**Comment by ivanhofer** [comment]
The question is: why do people know that there exists such a plugin? Where is this documented? We should just make it more clear that it is only meant to be used if users want to integrate it with a library that has no inlang plugin yet.

> * But the config schema should also not be a dedicated plugin itself

I disagree. If there is no dedicated plugin for a certain library, I won't be able to use the inlang ide extension

> * let this be not a plugin setting but a native vs code setting which the extension can expose

I don't think this is a good solution. Each person in the team will have to configure it manually if they don't commit the settings file. It is an extra hurdle to get started with the project, because you will need to figure out how to configure it.


**Summary**

I would just make it more clear when and how the plugin should be used.

**Comment by samuelstroschein** [comment]
@felixhaeberle:

> an ide-extension-plugin is needed – otherwise this can be set by any other plugin.*

That's the thing. It's not needed because a user can define the `ideExtension` prop themselves. The plugin doesn't nothing else. 

@ivanhofer:

>> But the config schema should also not be a dedicated plugin itself

> I disagree. If there is no dedicated plugin for a certain library, I won't be able to use the inlang ide extension

Maybe miscommunication. I was referring to the config schema of the ide extension should (likely) not be defined in the [plugin](https://github.com/inlang/inlang/blob/main/source-code/ide-extension-plugin/src/schema.ts) but the ide extension itself. Plugins should _define_ the config properties but not the _schema_ / the ide extension case differs a bit from the sdk-js-plugin because the sdk-js-plugin is required.  


>> let this be not a plugin setting but a native vs code setting which the extension can expose
> I don't think this is a good solution. Each person in the team will have to configure it manually if they don't commit the settings file. It is an extra hurdle to get started with the project, because you will need to figure out how to configure it.

@felixhaeberle seems to refer to exposing the config schema via the ide-extension itself for other plugins to define the config. Other plugins can't import other plugins / why should we have a plugin that defines the app config schema for an app and has no other purpose?

**Comment by felixhaeberle** [comment]
> That's the thing. It's not needed because a user can define the ideExtension prop themselves. The plugin doesn't nothing else.

Where can the user define this? I thought we decided on a plugin based architecture where such definitions are made by plugins and not for example on a global level, like it was before. 

> Other plugins can't import other plugins / why should we have a plugin that defines the app config schema for an app and has no other purpose?

Just wanted to have every option on the table. Most vs code extensions are working with native vs code settings which can be also easily set over the vs code extension API. This would benefit the user experience in a way that we don't introduce a new "settings pattern". 

> I don't think this is a good solution. Each person in the team will have to configure it manually if they don't commit the settings file. 

It's standard behavior to commit those settings in a .vscode folder when you want your team to have the same settings. 

Bottom line: I think the current state is good. I don't see a solution which is simple enough to be better then the current state (having a plugin which sets the settings when no library support is there) and with metaframework specific plugins setting this behavior, it's really easy to use the extension.

@ivanDo you see an easier & more user friendly solution?

**Comment by ivanhofer** [comment]
> It's standard behavior to commit those settings in a .vscode folder when you want your team to have the same settings.

But I think it will confuse people more if the `sdkPlugin` defines settings that don't show up in the settings page inside VS Code.

I think we should just slightly change the Docs in order to clear things and leave the rest as it is.

**Comment by felixhaeberle** [comment]
@ivanhofer Totally agree and especially: also write docs for the vs code extension. 

At the moment, we only have a README and while one README is usually enough, when the extension will envolve it could be beneficial to bring the users to our docs from a cross-selling perspective and not let all the traffic go to VS Marketplace where the README is displayed.

**Comment by samuelstroschein** [comment]
I close this issue as "the current state is okay".

**Comment by samuelstroschein** [comment]
FYI if we decide that my getting started proposal https://github.com/inlang/inlang/issues/755#issuecomment-1550131450 is the way to go, the ide extension plugin becomes obsolete because the ide extension is treated as app and thus should be in the core config of inlang

**Comment by samuelstroschein** [comment]
After fixing the ide extension and working with plugins in #845, I am confident that:

1. the ide extension config should be part of core 
2. the ide extension plugin is redundant. 

Plugins or users are expected to set the config. If the ide extension config is not part of core, the expectation is not "communicating", there is no typesafety, the config is not valided by `setupConfig` and last but not least the ide extension properties are relevant for inlang core in the future: 

- "Where is a message referenced in source code?"
- "How should messaged be extracted from source code?"

## Proposal 

1. Move the ide extension config to core under the `ideExtension` prop. In the future, we can step-by-step unbundle the `ideExtension` prop. 
2. Remove the `ide-extension-plugin`.

**Comment by ivanhofer** [comment]
> 1. the ide extension config should be part of core

What are the arguments in favor of this change and why would the sdk config not fit into the same category?

> ide extension properties are relevant for inlang core in the future:
> 
> * "Where is a message referenced in source code?"
> * "How should messaged be extracted from source code?"

I don't think it should be part of the IDE extension. A new module like [the proposed `adapters` here](https://github.com/pipeli18ne/RFC/blob/main/04-package-structure.md) would make sense.
In the future the sdk should be able to update code in files (e.g. rename key `welcome` to `welcome-message`) and bundling it into the IDE extension makes it harder to use it in other parts e.g. in the `cli`.

**Comment by samuelstroschein** [comment]
> What are the arguments in favor of this change and why would the sdk config not fit into the same category?

(Other) plugins are expected to set the ide extension config in contrast to the SDK. The SDK is not expected to have "plugins" setting the SDK settings. 

>>  "Where is a message referenced in source code?"
>> "How should messaged be extracted from source code?"

> I don't think it should be part of the IDE extension.

As mentioned, "In the future, we can step-by-step unbundle the ideExtension prop.". For now, it's simpler to put ide extension-related props under ide extension. In the future, we can "unbundle" those props if other apps require those props too.

**Comment by samuelstroschein** [comment]
After @jannesblobel tried to set up the IDE extension, I am even more certain that the dedicated ide extension plugin is wrong because it leads to massive confusion. 

## Recap on the inlang-plugin-ide-extension

The [ide extension plugin](https://github.com/inlang/inlang/tree/main/source-code/ide-extension-plugin) exists because https://github.com/inlang/inlang/issues/618 which had an incorrect assumption: 

> @inlang/core will become a massive config file if we allow non-shared config properties in the core config

By now we know that the IDE extension properties are shared across the inlang ecosystem. 

1. We expect plugins to set the config. 
2. The config properties provide information like "where is a message used in source code" which will be of interest across the inlang ecosystem (in the future). 

The ide extension plugin exists to define the ide extension config outside of @inlang/core. However, the plugin is redundant because it does not contain any logic by itself. (A fallback config is defined in the plugin but that pattern breaks and shouldn't be used.) 

## The problems with the ide extension plugin

### 1. The ide-extension plugins provides no value because it just defines a schema. The following two configs are identical: 

```js
// With ide-extension plugin

export async function defineConfig(env) {

  const { default: ideExtensionPlugin } = await env.$import(
    "https://cdn.jsdelivr.net/npm/@inlang/ide-extension-plugin@latest/dist/index.js"
  );

  return {
    referenceLanguage: "en",
    plugins: [
      ideExtensionPlugin({
        messageReferenceMatchers: undefined,
        // define property,
        extractMessageOptions: undefined,
        // define property
        documentSelectors: undefined,
        // define property
      }),
    ],
  };
}

```

```js
// Defining the config as part of inlang/core (which the ide extension plugin does under the hood too)

export async function defineConfig(env) {
  return {
    referenceLanguage: "en",
    ideExtension: {
      messageReferenceMatchers: undefined,
        // define property,
        extractMessageOptions: undefined,
        // define property
        documentSelectors: undefined,
        // define property
    }
  };
}
```

### 2. Users expect "plugins" to set the config. 

Thus, a correctly working ide extension config would be nothing more and nothing less than using a plugin like `i18next`, `typesafe-i18n-plugin`, `sdk-js-plugin` etc. 

```js
export async function defineConfig(env) {
  const { default: i18nextPlugin } = await env.$import(
    "https://cdn.jsdelivr.net/gh/samuelstroschein/inlang-plugin-json@latest/dist/index.js"
  );

  return {
    referenceLanguage: "en",
    plugins: [
      i18nextPlugin({
        pathPattern: "./path/to/languages/{language}.json",
      }),
    ],
  };
}
``` 

The whole concept of "configure the ide extension via a plugin" is extremely confusing if there is an "ide extension plugin" which people ofc think that's the one to go for while in reality its "nope, use another plugin".

**Comment by felixhaeberle** [comment]
@samuelstroschein So actually now we are moving from a everything in a plugin to "a bit in core and a bit in plugin" confusion as I follow along this thread. Correct me on this.

I said it once and I say it again: As a user, I expect the inlang plugin to just works. (— when I have inlang in use.) **Or doing i18n in general!**

Are we going a good path with "it works when you use one of our 1 metaframework? 2 file type? 3 i18n framework? 4 ide extension? specific plugins", but the plugin architecture here is so open, we have a hard time communicating what a plugin is, because they are 4 different communications about what a plugin is as I outlined above.

**Comment by samuelstroschein** [comment]
>So actually now we are moving from a everything in a plugin to "a bit in core and a bit in plugin" confusion as I follow along this thread. Correct me on this. 

@felixhaeberle no, i am proposing to move the config schema of the ide extension to `@inlang/core` and get rid of `ide-extension-plugin` because the latter is completely confusing as evident in this discussion. Nobody seems to understand how to set up the ide extension and what this plugin does (it does nothing, that's the problem). 

> I expect the inlang plugin to just works. 

Which plugin? There is no "the inlang plugin". 

I assume you refer to `ide-extension-plugin`. If you do, perfect illustration of my point. The ide extension plugin only defines the config schema for the ide extension, nothing else. Using `ideExtensionPlugin` doesn't make the ide extension work and neither can a generic `ideExtensionPlugin` be developed that works for every case. Thus, plugins like `i18next` should just set the ide extension config. 

> hard time communicating what a plugin is, 

A plugin sets inlang config properties. If a plugin sets ide extension properties, great the ide extension will work.

The `ide-extension-plugins` doesn't set any config properties because it defines additional properties outside of @inlang/core. Thus, nobody understands this plugin as it shouldn't exist in the first place. 


JESUS CHRIST HAHA THIS ISSUE IS SO TEDIOUS. MEET YA ALL IN AMSTERDAM and I explain my cause

**Comment by felixhaeberle** [comment]
@samuelstroschein But do you agree that we settle on one or max two plugin "types" to ease communication and handling of plugins? I also think that the ide extension plugin is not needed but we have to do this in a bigger scope and think which plugins we do not need too or hurt communication / handling.

**Comment by samuelstroschein** [comment]
@felixhaeberle yes. there is only one plugin type but the ide extension plugin suddenly led to a "second type of plugin" 😭

**Comment by felixhaeberle** [comment]
@samuelstroschein I still feel from a users perspective there are 3 types of plugins: 

one could think:

- ah, the plugins are **metaframework** based, when they use our **svelte plugin**.

- ah, the plugins are **file type** based, when they use our **json plugin**.

- ah, the plugins are **package** based, when they use our **i18n plugin**.

which is confusing. and to not confuse people, great communication is needed here.

**Comment by samuelstroschein** [comment]
> which is confusing. and to not confuse people, great communication is needed here.

All your plugin examples boil down to: "What is your stack?"

> - ah, the plugins are metaframework based, when they use our svelte plugin.

Your stack is svelte, we recommend our svelte SDK that has a plugin.

> - ah, the plugins are file type based, when they use our json plugin.

You use a stack that has no dedicated plugin but your resources are json, you can use the json plugin. _Important to mention: The JSON plugin is a fallback. It was never meant to become the most used plugin_.
 
> - ah, the plugins are package based, when they use our i18n plugin.

Your stack is framework X and you use an i18n library that has a plugin, use `inlang-plugin-{framework}`. 

---

The upcoming docs revamp is supposed to clarify this situation by having a plugin registry which is heavily stack focused. Basically, search for your stack, done. See https://github.com/inlang/inlang/issues/778

**Comment by felixhaeberle** [comment]
@NilsJacobsen @NiklasBuchfink I think this is really important to have it cleary described in the docs, for me the user probably the most "confusing part" if he can not make this resolution of plugin choice, he is lost or might misconfigure the config. 

I'm so happy we finally will get this sorted out in the docs 🫶🎉

**Comment by felixhaeberle** [comment]
@samuelstroschein can we close this in favor of a new issue with all steps required to deprectate the ide extension plugin?

**Comment by samuelstroschein** [comment]
@felixhaeberle do we agree on removing the ide extension plugin? 

If so, I will create a follow up issue and get to work.

**Comment by samuelstroschein** [comment]
Closing this issue as "ide extension plugin will be removed, and the config be moved to core". Follow up progress is tracked in #856
